[Elliot Lim Zhi Yong] iP

Spyobird opened the pull request:
# Juke
> A simple to use task tracker.

Features:
- Handles three tasks types (todo, deadline and event)
- Saves data locally
- Easy to use CLI

How to install:
1. Download the jar from [here](https://github.com/Spyobird/ip/releases).
2. Move it to the intended directory.
3. Double click it or run `java -jar Juke.jar` in the command line.
4. Tada! 😄

Simple to use and completely **FREE**!
***
Future implementations:
- [ ] GUI
- [ ] More task types

Here is the `main` method.
```java
public static void main(String[] args) {
    INSTANCE.run();
}
```

Commit:
Add Gradle support

Spyobird committed:
Rename Duke.java to Juke.java; add greet, echo and exit functionality to Juke.java

Spyobird committed:
Juke.java: add functionality for add, display task list

Spyobird committed:
Add Task.java; Add TaskStatus.java

Spyobird committed:
Juke.java: Add token/delimitting of input; Add mark/unmark for task list

Spyobird committed:
Add Deadline.java; Add Event.java; Add Todo.java

Spyobird committed:
Task.java: Made abstract

Spyobird committed:
Juke.java: Add functionality for more types of tasks (todo, deadline, events)

Spyobird committed:
Update I/O redirection for semi-auto testing

Spyobird committed:
Add JukeException.java; Add sub-classes of JukeException.java

Spyobird committed:
Move Task.java, Deadline.java, Event.java, TaskStatus.java, Todo.java to the task package

Spyobird committed:
Juke.java: Add basic error handling

Spyobird committed:
Juke.java: Add delete functionality, Replace array with collection

Spyobird committed:
Add FileManager.java; Add PrintHelper.java

Spyobird committed:
Juke.java: Add file save functionality; Task.java: Add getStatus method

Spyobird committed:
Add directory for juke.txt config file

Spyobird committed:
Shift files to new package

Spyobird committed:
Add DateTimeHandler.java

Spyobird committed:
Merge branch 'branch-Level-7'

Spyobird committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/juke/Juke.java

Spyobird committed:
Add Command.java, CommandHandler.java, EchoCommand.java, Result.java

Spyobird committed:
Add Parser.java, Storage.java, TaskList.java, Ui.java

Spyobird committed:
Juke.java: Add echo as test command

Spyobird committed:
Add JavaDoc

Spyobird committed:
Add remainder of JavaDoc

Spyobird committed:
Change variable declaration to fit coding standard

Spyobird committed:
Add find function

Spyobird committed:
Merge branch 'branch-A-JavaDoc'

Spyobird committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/juke/command/Command.java
#	src/main/java/juke/task/Task.java

Spyobird committed:
Task.java: Remove accidental explicit class declaration

Spyobird committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/juke/common/TaskList.java

Spyobird committed:
Tidy up Result.java and associated classes

Spyobird committed:
Add subclasses for Command.java; Juke.java: Update with new command usage

Spyobird committed:
Juke.java: Simplify switch cases to a simple UI loop; Ui.java: Create UI loop

Spyobird committed:
Add storage functionailty, clean up rough code; Juke.java: Clean up running loop; Storage.java: Handle storage functions

Spyobird committed:
Delete FileManager.java and PrintHelper.java which are replaced by Ui.java and Storage.java

Spyobird committed:
Storage.java: Fix a small bug with Juke instances

Spyobird committed:
Add tests for Parser.java and Ui.java

Spyobird committed:
Add ParserTest.java (missed on last commit)

sibinhho99 commented on src/main/java/juke/Juke.java:
```diff
@@ -0,0 +1,271 @@
+        this.hasExited = true;
+    }
+    /*
+    private void echo(String[] args) {
```
Perhaps you should delete this

sibinhho99 commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,114 @@
+
+import java.time.format.DateTimeParseException;
+
+public class AddCommand extends Command {
```
Perhaps you should add header comment in this method

sibinhho99 commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,114 @@
+public class AddCommand extends Command {
+    private TaskType type;
+    
+    public AddCommand(TaskType type) {
```
Perhaps you should add header comment

sibinhho99 commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,114 @@
+        return this;
+    }
+    
+    @Override
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,114 @@
+        return this;
+    }
+    
+    public TaskType getType() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+    }
+    
+    @Override
+    public Command execute() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/EchoCommand.java:
```diff
@@ -0,0 +1,39 @@
+    private String message;
+    
+    @Override
+    public Command checkParametersAndArguments() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/ExitCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+public class ExitCommand extends Command {
+    @Override
+    public Command checkParametersAndArguments() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/ExitCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+import juke.exception.JukeException;
+
+public class ExitCommand extends Command {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/ExitCommand.java:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    @Override
+    public Command execute() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/FindCommand.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.List;
+
+public class FindCommand extends Command {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/FindCommand.java:
```diff
@@ -0,0 +1,45 @@
+
+public class FindCommand extends Command {
+    @Override
+    public Command checkParametersAndArguments() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/FindCommand.java:
```diff
@@ -0,0 +1,45 @@
+    }
+    
+    @Override
+    public Command execute() {
```
Perhaps you should add header comment here

sibinhho99 commented on src/main/java/juke/command/ListCommand.java:
```diff
@@ -0,0 +1,41 @@
+import juke.exception.JukeException;
+import juke.exception.JukeTaskListEmptyException;
+
+public class ListCommand extends Command {
```
Perhaps you should add header comment here and in public methods

sibinhho99 commented on src/main/java/juke/command/MarkCommand.java:
```diff
@@ -0,0 +1,68 @@
+import juke.exception.JukeException;
+import juke.task.TaskStatus;
+
+public class MarkCommand extends Command {
```
Perhaps you should add header comment here and in public methods

sibinhho99 commented on src/main/java/juke/common/Storage.java:
```diff
@@ -0,0 +1,172 @@
+        }
+    }
+    
+    public void loadTasks() {
```
Perhaps you should add header comment here and in other public methods

sibinhho99 commented on src/main/java/juke/common/Parser.java:
```diff
@@ -0,0 +1,41 @@
+     * @return List of string arrays with the above specifications.
+     */
+    public static ArrayList<String[]> parseInput(String input) {
+        String[] paramSplit = input.strip().split(" -");
```
This variable name could be more descriptive

sibinhho99 commented on src/main/java/juke/exception/JukeInvalidArgumentCountException.java:
```diff
@@ -0,0 +1,18 @@
+    private int expected;
+    private int count;
+    
+    public JukeInvalidArgumentCountException(String cmd, int expected, int count) {
```
Perhaps you could add header comment here

sibinhho99 commented on src/main/java/juke/exception/JukeInvalidArgumentCountException.java:
```diff
@@ -0,0 +1,18 @@
+    }
+    
+    @Override
+    public String getMessage() {
```
Perhaps you could add header comment here

zhengtaoJ commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,114 @@
+    public AddCommand(TaskType type) {
+        this.type = type;
+    }
+    
```
Consider adding a header comment for this method?

zhengtaoJ commented on src/main/java/juke/command/Command.java:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return Boolean result.
+     */
+    public boolean isErranous() {
```
I believe it's a typo here?

ianfromdover replied:
```suggestion
    public boolean isErroneous() {
```
Small typo :)

zhengtaoJ commented on src/main/java/juke/command/EchoCommand.java:
```diff
@@ -0,0 +1,39 @@
+    }
+    
+    @Override
+    public Command execute() {
```
Consider adding header comments here?

zhengtaoJ commented on src/main/java/juke/DateTimeHandler.java:
```diff
@@ -0,0 +1,43 @@
+import java.time.format.DateTimeParseException;
+import java.util.Locale;
+
+public class DateTimeHandler {
```
Consider adding header comments here?

Spyobird committed:
Delete unneeded exceptions; Add more descriptive responses

Spyobird committed:
Add exception handling for AddCommand.java with regards to date and time parsing

ianfromdover commented on src/main/java/juke/command/Command.java:
```diff
@@ -0,0 +1,204 @@
+        if (this.hasParameter(param)) {
+            return this.paramArgs.get(param).orElseGet(() -> {
+                this.result = Result.error(new JukeInvalidParameterException(param));
+                return "";
```
Good use of streams here 👍

ianfromdover commented on src/main/java/juke/command/CommandHandler.java:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * Initializes the commands.
+     */
+    public static void registerCommands() {
+        COMMANDS.put("echo", () -> new EchoCommand());
+        COMMANDS.put("bye", () -> new ExitCommand());
+        COMMANDS.put("list", () -> new ListCommand());
+        EnumSet.allOf(TaskStatus.class)
+                .forEach(status -> COMMANDS.put(status.getCommandName(), () -> new MarkCommand(status)));
+        EnumSet.allOf(TaskType.class)
+                .forEach(type -> COMMANDS.put(type.getCommandName(), () -> new AddCommand(type)));
+        COMMANDS.put("delete", () -> new DeleteCommand());
+        COMMANDS.put("find", () -> new FindCommand());
+    }
```
Very well organised and easy to understand 👍

ianfromdover commented on src/main/java/juke/common/Ui.java:
```diff
@@ -0,0 +1,124 @@
+    private static final String LINE_DIVIDER = "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            + "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";
+    private static final String LINE_INDENT = "    ";
+    private static final String LOGO = "                                                             _____   \n"
+        + "          _____  ______   _____     ______   _______    _____\\    \\  \n"
+        + "         |\\    \\_\\     \\  \\    \\   |\\     \\  \\      \\  /    / |    | \n"
+        + "         \\ \\     \\\\    |  |    |    \\\\     \\  |     /|/    /  /___/| \n"
+        + "          \\|      ||   |  |    |     \\|     |/     //|    |__ |___|/ \n"
+        + "           |      ||    \\_/   /|      |     |_____// |       \\       \n"
+        + "   ______  |      ||\\         \\|      |     |\\     \\ |     __/ __    \n"
+        + "  /     / /      /|| \\         \\__   /     /|\\|     ||\\    \\  /  \\   \n"
+        + " |      |/______/ | \\ \\_____/\\    \\ /_____/ |/_____/|| \\____\\/    |  \n"
+        + " |\\_____\\      | /   \\ |    |/___/||     | / |    | || |    |____/|  \n"
+        + " | |     |_____|/     \\|____|   | ||_____|/  |____|/  \\|____|   | |  \n"
+        + "  \\|_____|                  |___|/                          |___|/   \n";
```
Good placement of `+` operator 👍

ianfromdover reviewed (commented):
Well written and follows coding standards closely 👍 Keep up the good work!

lovemathboy commented on src/main/java/juke/Juke.java:
```diff
@@ -0,0 +1,89 @@
+            this.ui.runUiLoop();
+        }
+    }
+    
```
I like how you abstracted out everything to their constituent classes! Makes code significantly easier to read!

lovemathboy commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,112 @@
+            } else {
+                this.result = Result.error(new JukeMissingArgumentException(this.type.getCommandName()));
+                return this;
+            }
```
Wouldn't it be better to rearrange the if statement so that 1: it avoids arrowhead coding and 2: there's no blank if statement?
```
            if (!this.hasParameter("at")) {
                this.result = Result.error(new JukeMissingArgumentException(this.type.getCommandName()));
                return this;
            } else if (!this.hasArgument("at")) {
                this.result = Result.error(new JukeMissingArgumentException(this.type.getCommandName()));
                return this;
            }
```

lovemathboy commented on src/main/java/juke/command/AddCommand.java:
```diff
@@ -0,0 +1,112 @@
+            return this;
+        }
+        return this;
+    }
```
"Result.error(new JukeMissingArgumentException(this.type.getCommandName()))" is repeated a lot, consider create a new method to fetch the result with only "this.type.getCommandName()" as the input?

lovemathboy reviewed (commented):
I especially liked the use of an abstract class for each type of command! Well done!

Spyobird committed:
Storage.java: Change spelling

Spyobird committed:
Merge branch 'add-gradle-support'

Spyobird committed:
build.gradle: Change main class name

Spyobird committed:
Fix a typo with "errorneous"

Spyobird committed:
build.gradle: Update main class name for shadowJar; Result.java: Type cast to String[] to suppress warnings

Spyobird committed:
Merge branch 'branch-A-Gradle'

Spyobird committed:
Reformat for checkstyle errors

Spyobird committed:
Merge branch 'branch-A-Checkstyle'

Spyobird committed:
Add basic JavaFX setup

Spyobird committed:
Adjust characters for better CLI support

Spyobird committed:
Move and rename Ui.java to TextUi.java; Rename UiTest.java to TextUiTest.java; Create Gui.java; Initialize basic components for the JavaFX UI

Spyobird committed:
Build GUI using FXML

There are two controller classes for the GUI, namely DialogBox and
MainWindow. They have their respective Java classes alongside their
FXML declarations.

Edited CommandHandler.java to include greater abstraction for
fetching commands and results for use with TextUi.java and GUI.java.

Renamed and modified TextUiTest.java to CommandHandlerTest.java to
reflect the refactoring of methods.

GUI so far is basic but functional.

Spyobird committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/juke/ui/TextUi.java

Spyobird committed:
Add header comments to all methods and classes.

Spyobird committed:
Add simple assertions

Spyobird committed:
Merge pull request #2 from Spyobird/branch-A-Assertions

Merge branch-A-Assertions

Spyobird committed:
Add additional cases for time and date parsing

DateTimeHandler.java now contains additional strings for various
configurations of date and time, so as to minimize the possibility
of incorrect parsing. Optional and default values are also added
in the event of incomplete parsing.

Spyobird committed:
Merge pull request #3 from Spyobird/branch-A-Streams

Add additional cases for time and date parsing

Spyobird committed:
Merge branch 'master' into branch-A-CodeQuality

Spyobird committed:
Add clone command

Command allows users to clone existing tasks in the task list.
Implements Java's Cloneable interface, creating deep copies of
task instances through the implementation of clone(). The cloned
tasks are then added to the end of the task list.

Spyobird committed:
DateTimeHandler.java: Fix severe bug for date time parsing

Spyobird committed:
Add TimeTask.java as an abstraction for tasks with a time element

Spyobird committed:
Add edit command

Spyobird committed:
Merge branch 'branch-C-Update'

Spyobird committed:
Merge branch 'master' into branch-A-CodeQuality

Spyobird committed:
Refactor code for better quality and readability

Spyobird committed:
Add code refactoring to follow better code quality

Large scale refactoring of subclasses of Command.java to allow
better readability and more intuitive understanding.
Removal of many instance of unnecessary uses of 'this'.
Refactoring of some parsing code to Parser.java.
Added more intuitive and natural error handling and propagation to
methods in TaskList.java.

Spyobird committed:
Merge branch 'branch-A-CodeQuality'

Spyobird committed:
Edit checkstyle.xml to ignore incorrect warnings

Type-cast and lambdas has incorrect checkstyle errors that failed
during build.

Spyobird committed:
Edit build.gradle to fix a severe error with building JavaFX

Spyobird committed:
Add better management of the running instance of Juke.

Spyobird committed:
Change profile icons for GUI

Spyobird committed:
Add Ui.png

Spyobird committed:
Fix severe error with find command

Spyobird committed:
Update README.md for user guide

Spyobird committed:
Update README.md to reflect correct time and date handling

Spyobird committed:
Add stage title for Juke

Spyobird committed:
Update README.md for small changes

Spyobird committed:
Set theme jekyll-theme-cayman

Spyobird committed:
Update _config.yml

Testing if Github Pages works

Spyobird committed:
Update _config.yml

Try out new Github Themes with Jekyll.

Spyobird committed:
Update _config.yml

Revert theme as it was not working.

Spyobird committed:
Update README.md

Spyobird committed:
Merge branch 'master' of https://github.com/Spyobird/ip

Spyobird committed:
Update README.md

Spyobird committed:
Update _config.yml

Add a fullstop.

Spyobird committed:
Update GUI to handle resizability

Spyobird committed:
Fix minor bug with text wrapping for JavaFX labels

Spyobird committed:
Add sample data population on new file creation

Spyobird committed:
Update Ui.png

Spyobird committed:
Add delete all functionality

Spyobird committed:
Attempt to fix occasional thread blocking on bye command

Calls to Thread.sleep() causes blocking effects on JavaFX
application thread. It appears that adding the Juke response to
the bye command is not thread safe, and occasionally the
application exits without showing the bye command due to the
thread block. Attempted to use JavaFX concurrency, however, the
problem still persists albeit at a lower frequency.
